semgrep/semgrep/rule_match.py

```diff
@@ -324,8 +359,11 @@ class RuleMatchSet(Set[RuleMatch]):
     It also automagically adds the correct finding index when adding elements.
     """
 
-    def __init__(self, __iterable: Optional[Iterable[RuleMatch]] = None) -> None:
+    def __init__(
+        self, rule: Rule, __iterable: Optional[Iterable[RuleMatch]] = None
```

Comment:
Hmm I'm a little worried about associating a Rule to a RuleMatchSet. On one hand it makes sense that deduplication only matters for matches with the same rule anyway but it was nice that this data-structure worked even if the matches you put in it were from different rules.

Should a RuleMatch just maintain a reference to the Rule it matched with so we can create the match_formula_string without keeping state in RuleMatchSet. Thoughts @underyx?

Reply:
Looks like @ajbt200128 you had the same idea as me: https://github.com/returntocorp/semgrep/pull/5470/files#diff-235785e37f4fac53689eae23436d86080f6e09721a4904ed674be882e4f38505R80-R81

Reply:
Yea, I figured this made sense since "set" means deduping, and I'd actually disagree that it's nice it work across different rules, I actually think that doesn't make sense. If we have to dedup across rules that doesn't make sense to me. If two rules have two matches that are considered duplicates, wouldn't the rules be duplicates? As far as just keeping a rule reference, we can do that but we still would rely on RuleMatchSet for the index, which the new hash needs, so it felt safer to couple the logic together.

Reply:
it's less that we would use the datastructure to dedup findings that were duplicates and more that: before associating a rule to the rulematchset, the rulematchset could contain matches from any rule and them being not duplicates would just work.

Reply:
Actually can we add a check to make sure we are not adding a rulematch whos rule_id doesnt match the RuleMatchSet's rule

Reply:
I don't fully grasp this change but I will say that this class is supposed to be a native Python `set` just with extra features, so we shouldn't change the function signature from the original `def __init__(self, __iterable)`. If we need to add more stuff, I'd replace this with a type that keeps data in a set in an instance variable.

Reply:
Will do that